README.md
# godex-printer

Sends EZPL print commands to GoDex label printer via serial port. Uses SVG as template for labels. For more information on GoDex's EZPL go [here](http://www.godexintl.com/EN/download/downloads/list/Manuals).

---

#### __Installation__

```bash
npm install godex-printer
```

---

#### __Quick Start__

```javascript
// Import necessary modules
import {Printer, Label} from 'godex-printer';

var printer = new Printer({port:'COM4'}); // Create a new printer

// All units are in mm
var label1 = new Label({width: 80, height: 52}); // Create a new label.
label1.addRect(0,0,73.5,49,0.1);
label1.addLineHor(0, 73.5, 24, 0.2);
label1.addLineVer(35, 24, 49, 0.2);
label1.addText("Part Number", 2, 1, 2);
label1.addText("LREM2019", 1, 4, 10);
label1.addBarcode('CODE39', 2, 14, 0.2, 0.75, 8, "7dds18891");
label1.addText("QTY (PCS)", 2, 25, 2);
label1.addText("19", 2, 28, 6);
label1.addText("BIN", 36, 25, 2);
label1.addText("Tray", 36, 28, 6);

// Deprecated -> printer.addPrintTask(label1.getPrintCommand());
printer.printLabel(label1);

printer.on("printQueueEmpty", function(){
   console.log("Everything printed");
   printer.stop();
});
```

___Note:___ All units are in dots (dot as in dpi. For 203 dpi printer 1mm = 8 dots)

---

#### __Label template in SVG__

```javascript
import {Printer, SvgLabel} from 'godex-printer';

var printer = new Printer();
printer.start('COM4')
.then(function(){
   var svgLabel = new SvgLabel(__dirname+'/L01.svg', {'PartNo':7188, 'uom': 'pcs', 'Qty': 122, 'Bin': 'BXR109'});
   // Deprecated -> printer.addPrintTask(svgLabel.getPrintCommand());
   printer.printLabel(svgLabel);
   printer.on("printQueueEmpty", function(){
      console.log("Everything printed");
      printer.stop();
   });
});
```

SVG template file contains:

```svg
<svg width="80mm" height="52mm" style="border: 1px solid #CCC" data-template-name="test" labelGap="2" leftMargin= "26" rowOffset= "-15" startPos= "20">
   <!-- Rectangle -->
   <rect x="0mm" y="0mm" width="73.5mm" height="49mm" style="stroke: #000000; fill: none;" stroke-width="0.5mm" />
   <!-- Line -->
   <line x1="0mm" y1="24mm" x2="73.5mm" y2="24mm" stroke="red" stroke-width="0.2mm" fill="none" />
   <!-- Line -->
   <line x1="35mm" y1="24mm" x2="35mm" y2="49mm" stroke="red" stroke-width="0.2mm" fill="none" />
   <text x="2mm" y="3mm" font-size= "2mm">PART NUMBER</text>
   <text x="2mm" y="10mm" font-size= "6mm">{PartNo}</text>
   <text x="2mm" y="22mm" font-size= "8mm" data-barcode-type='CODE39'>#BA#-PartNo</text>
   <text x="2mm" y="27mm" font-size= "2mm">QTY ({uom})</text>
   <text x="2mm" y="34mm" font-size= "6mm">{Qty}</text>
   <text x="36mm" y="27mm" font-size= "2mm">BIN</text>
   <text x="36mm" y="34mm" font-size= "6mm">{Bin}</text>
   <text x="55mm" y="45mm" font-size= "2mm">#DATE#</text>
</svg>
```

___Note:___ All units in SVG must be in mm (For 203 dpi printer 1mm = 8 dots). Exceptions are: _leftMargin_, _rowOffset_ and _startPos_ which are in dots.

---

#### __Label Properties__

| Property  | Description | Default Value |
| ------- | ---------------- | ---------------- |
| speed | Print speed | 4 |
| darkness | Print ink darkness | 26 |
| roate | Label rotation | 150 |
| leftMargin | Margin to leave from left of lable | 26 |
| rowOffset | Margin to leave from top of label | -15 |
| width | Width of label | 80 (in mm) |
| length | Length of label | 52 (in mm) |
| Gap | Gap between labels | 2 (in mm) |
| startPos | Label feed position after printing | 20 |
| copies | Number of copies of a label to print | 1 |

---

### __License__

MIT


history.md
1.1.0
==================

* Changed Label property length to height. Now use new Label({height: 40})
* All input values are in millimeters
* Moved to using promises

1.0.2 / 2016-09-06
==================

* Added build script

1.0.1 / 2016-09-05
==================

* Added SVG label template feature and examples.

1.0.0 / 2016-09-05
==================

* Initial release


src/svg2ezpl.js
/*jshint esversion: 6 */

/*
svg2ezpl.js
===========
A module to reade SVG and spit out EZPL command for GoDex printers.
*/
import fs from 'fs';
import parse from 'xml-parser';
import _ from 'underscore';
import moment from 'moment';

export function svg2ezpl(){

   this.toDot = 8;
   this.labelWidth = 70;
   this.labelHeight = 30;
   /*
   * Get print command for given data and template
   */
   this.getPrintCmd= function(data, template){
      // Read SVG file and parse into JSON
      var xml = fs.readFileSync('app/templates/' + template + '.svg', 'utf8');
      var obj = parse(xml);
      var result = "";
      _.each(obj.root.children, function(child) {
         var att = child.attributes;
         var name = child.name;

         // remove unit - mm
         att = _.mapObject(att, function(val, key) {
            return val.replace("mm", "");
         });

         if(name == "line"){
            result += this.getLine(att);
         }
         else if(name == "rect"){
            result += this.getRect(att);
         }
         else if(name == "text"){
            // Barcode
            if(child.content.indexOf("#BA#") >= 0)
               result += this.getBarCode(child.content, att, data);
            // Date
            else if(child.content.indexOf("#DATE#") >= 0){
               result += this.getText(moment().format('DD-MMM-YY'), att, data);
            }
            // Normal Text
            else
               result += this.getText(child.content, att, data);
         }
      }, this);
      return "^L\n" + result + "E\n" ;
   };

   /*
   * Convert SVG Line to EZPL Line
   */
   this.getLine = function(att){
      var x2, y2;
      if(att.x1 == att.x2){
         x2 = (1*att.x2) + (1*att['stroke-width']);
         y2 = att.y2;
      }

      else if(att.y1 == att.y2){
         x2 = att.x2;
         y2 = (1*att.y2) + (1*att['stroke-width']);
      }

      var line = "La," + (att.x1*this.toDot) + "," + (att.y1*this.toDot) + ","  + x2*this.toDot + "," + y2*this.toDot + "\n";
      return line;
   };

   /*
   * Convert SVG Rectangle to EZPL Rectangle
   */
   this.getRect = function(att){
      var x1 = att.x * this.toDot;
      var y1 = att.y * this.toDot;
      var x2 = ((att.x*1) + (att.width*1)) * this.toDot;
      var y2 = ((att.y*1) + (att.height*1)) * this.toDot;
      var w = Math.floor(att['stroke-width'] * this.toDot);
      var line = "R" + x1 + "," + y1 + "," + x2 + "," + y2 + "," + w + "," + w + "\n";
      return line;
   };

   /*
   * Convert SVG Text to EZPL Text
   */
   this.getText = function(content, att, data){
      var line = "";
      var val = content;
      _.each(_.keys(data), function(key){
         val = val.replace("{"+ key +"}", data[key]);
      });
      var fs = att["font-size"]*1;
      var x = att.x * this.toDot;
      var y = ((att.y*1) - fs) * this.toDot;
      var w = att["font-size"]*this.toDot;
      var b = att["stroke"]? "0B" : "0";
      if(b == "0"){
         var i = [6, 8, 10, 12, 14, 18, 24, 30];
         var j = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
         fs = Math.floor(fs * 3.75);
         var k = _.map(i, function(num){ return fs>num?(fs-num):(num-fs); });
         var l = _.lastIndexOf(k, _.min(k));
         line = "A" + j[l] + "," + x + "," + y + ",1,1,1,0,"+ val + "\n";
      }
      else
      line = "AT," + x + "," + y + "," + w + "," + w  + ",1," + b + ",0,0," + val + "\n";
      return line;
   };

   // Convert SVG Text to EZPL  Barcode
   this.getBarCode = function(content, att, data){
      var val = content;
      _.each(_.keys(data), function(key){
         val = val.replace("#BA#-"+ key, data[key]);
      });
      var line = "BA,";
      var x = att.x * this.toDot;
      var fs = att["font-size"]*1;
      var y = ((att.y*1) - fs) * this.toDot;

      line = line + x + "," + y + ",2,4,"+fs*this.toDot+",0,0," + val + "\n";
      return line;
   };
}
